File: 22-generate-parentheses/generate-parentheses.java
class Solution {
    StringBuilder temp;
    List<String> ans;

    void helper(int open,int close,int n)
    {
        if(open == n && close == n)
        {
            ans.add(temp.toString());
            return;
        }

        if(open < n)
        {
            temp.append('(');
            helper(open+1,close,n);
            temp.deleteCharAt(temp.length() - 1);
        }

        if(close < open)
        {
            temp.append(')');
            helper(open,close+1,n);
            temp.deleteCharAt(temp.length() - 1);
        }
    }

    public List<String> generateParenthesis(int n) {
        temp = new StringBuilder();
        ans = new ArrayList<String>();
        helper(0,0,n);
        return ans;
    }
}

File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
public:

    void helper(int open,int close,string &temp,vector<string> &ans, int n)
    {
        if(open == n && close == n)
        {
            ans.push_back(temp);
            return;
        }

        if(open < n)
        {
            temp.push_back('(');
            helper(open+1,close,temp,ans,n);
            temp.pop_back();

        }
        if(close < open)
        {
            temp.push_back(')');
            helper(open,close+1,temp,ans,n);
            temp.pop_back();
        }
    }

    vector<string> generateParenthesis(int n) {
        string temp = "";
        vector<string> ans;

        helper(0,0,temp,ans,n);
        return ans;
    }
};